crates/core/component/auction/src/component/dutch_auction.rs

```diff
@@ -51,11 +64,183 @@ pub(crate) trait DutchAuctionManager: StateWrite {
+    /// credit and zero-out its reserves, and finally, examine the auction's termination
+    /// condition.
+    async fn execute_dutch_auction(
+        &mut self,
+        auction_id: AuctionId,
+        trigger_height: u64,
+    ) -> Result<()> {
+        let old_dutch_auction = self
+            .get_dutch_auction_by_id(auction_id)
+            .await
+            .expect("no deserialization errors")
+            .expect("the auction exists");
+
+        let DutchAuctionDescription {
+            input,
+            output_id,
+            max_output: _,
+            min_output: _,
+            start_height,
+            end_height,
+            step_count,
+            nonce: _,
+        } = old_dutch_auction.description;
+
+        let DutchAuctionState {
+            sequence: _,
+            current_position,
+            next_trigger: _,
+            input_reserves: _,
+            output_reserves: _,
+        } = old_dutch_auction.state;
+
+        let auction_input_id = input.asset_id;
+        let auction_output_id = output_id;
+
+        let auction_trigger = TriggerData {
+            start_height,
+            end_height,
+            step_count,
+        };
+
+        // Recover the LP's balances, if it exists.
+        let lp_reserves = if let Some(auction_lp_id) = current_position {
+            self.close_position_by_id(&auction_lp_id)
+                .await
+                .map_err(|e| {
+                    tracing::error!(
+                        ?e,
+                        ?auction_lp_id,
+                        ?auction_id,
+                        "failed to close dutch auction LP"
+                    )
+                })
+                .expect("position should exist and be opened or closed");
+            self.withdraw_position(auction_lp_id, 0u64)
+                .await
+                .map_err(|e| {
+                    tracing::error!(
+                        ?e,
+                        ?auction_lp_id,
+                        ?auction_id,
+                        "failed to close dutch auction LP"
+                    )
+                })
+                .expect("no state incoherence")
+        } else {
+            Balance::zero()
+        };
+
+        // We remove the execution trigger that we are currently processing:
+        self.unset_trigger_for_dutch_id(auction_id, trigger_height);
+
+        // Prepare a new auction, based on the previous one.
+        let mut new_dutch_auction = DutchAuction {
+            description: old_dutch_auction.description,
+            state: old_dutch_auction.state,
+        };
+
+        let balance_input_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == input.asset_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
+        let balance_output_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == output_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
```

Comment:
This feels a bit dangerous. If for some reason the asset IDs were mismatched, we wouldn't notice and we'd just pass through a zero, right? Or do we have to do that because of the way `Balance` works (does it record zero entries?)

Reply:
We're getting this `Balance` from withdrawing the LP, but would adding a check that:
- there are no negative entries
- there are all from the correct asset ids
alleviate the concern here?

Reply:
I'm not sure it's an important concern, if the `Balance` isn't creating zero entries it may not have entries for our assets anyways.